File: web/src/wasm/demo/demoData.ts
// TS port of demo_data.py: the seeded synthetic series + the asset geometries.
// Numpy's RNG can't be reproduced bit-for-bit, so we use a small seeded PRNG;
// the *shapes* (lulls/gusts, daily solar bell, battery cycle, forecast fan-out)
// match. 72 hourly points from 2026-01-01. Timestamps are naive-UTC ISO strings
// ("YYYY-MM-DDTHH:MM:SS"), matching the strftime format the read queries emit.

export const PERIODS = 72;
const START_MS = Date.UTC(2026, 0, 1, 0, 0, 0);

export interface Point {
  t: string; // valid_time, ISO naive-UTC
  v: number; // value
}

export function isoHour(i: number): string {
  return new Date(START_MS + i * 3_600_000).toISOString().slice(0, 19);
}

// --- seeded PRNG (mulberry32) + gaussian -----------------------------------
function rng(seed: number): () => number {
  let s = seed >>> 0;
  return () => {
    s = (s + 0x6d2b79f5) >>> 0;
    let t = s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}
function gauss(r: () => number, mu = 0, sigma = 1): number {
  const u = Math.max(r(), 1e-12);
  const v = r();
  return mu + sigma * Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}
const clip = (x: number, lo: number, hi: number) => Math.min(Math.max(x, lo), hi);
const round3 = (x: number) => Math.round(x * 1000) / 1000;
const hod = (i: number) => i % 24;
const frame = (vals: number[]): Point[] => vals.map((v, i) => ({ t: isoHour(i), v: round3(v) }));

// --- capacity-factor random walk (lulls + gusts) ---------------------------
function windCf(seed = 2): number[] {
  const r = rng(seed);
  const cf = new Array<number>(PERIODS);
  cf[0] = 0.5;
  for (let i = 1; i < PERIODS; i++) {
    cf[i] = clip(0.92 * cf[i - 1] + 0.08 * 0.5 + gauss(r, 0, 0.025), 0, 1);
  }
  return cf;
}

export const windPower = (capacity = 3.5, seed = 2): Point[] =>
  frame(windCf(seed).map((c) => capacity * c));

export const windSpeed = (seed = 2): Point[] =>
  frame(windCf(seed).map((c) => 3.0 + 9.0 * Math.cbrt(c)));

export function solarPower(capacity = 10, seed = 1): Point[] {
  const r = rng(seed);
  return frame(
    Array.from({ length: PERIODS }, (_, i) => {
      const bell = clip(Math.sin((Math.PI * (hod(i) - 6)) / 12), 0, Infinity);
      const clouds = 0.8 + 0.2 * r();
      return clip(capacity * bell * clouds, 0, capacity);
    }),
  );
}

function batterySocVals(seed = 3): number[] {
  const r = rng(seed);
  return Array.from({ length: PERIODS }, (_, i) => {
    const soc = 55 + 30 * Math.sin((2 * Math.PI * (hod(i) - 15)) / 24);
    return clip(soc + gauss(r, 0, 1.5), 5, 99);
  });
}
export const batterySoc = (seed = 3): Point[] => frame(batterySocVals(seed));

export function batteryPower(seed = 3): Point[] {
  const soc = batterySocVals(seed);
  return frame(soc.map((s, i) => (i === 0 ? 0 : (s - soc[i - 1]) * 5.0)));
}

// --- forecast revisions: bounded windows that start + end inside the timeline
export const FORECAST_HORIZON_H = 36;
export const FORECAST_REVISIONS: [number, number][] = [
  [0, 0.34],
  [8, 0.27],
  [16, 0.2],
  [24, 0.14],
  [32, 0.09],
  [40, 0.05],
];

export function windPowerForecast(
  capacity = 3.5,
  issuedAtH = 0,
  error = 0.15,
  horizonH = FORECAST_HORIZON_H,
  seed = 2,
): Point[] {
  const truth = windCf(seed).map((c) => capacity * c);
  const end = Math.min(issuedAtH + horizonH, PERIODS);
  const r = rng(200 + issuedAtH);
  const phase = r() * 2 * Math.PI;
  const sign = r() < 0.5 ? 1 : -1;
  const out: Point[] = [];
  for (let idx = issuedAtH; idx < end; idx++) {
    const lead = idx - issuedAtH;
    const growth = lead / Math.max(horizonH, 1);
    const wave = Math.sin((2 * Math.PI * lead) / 52.0 + phase);
    const dev = sign * error * growth * wave;
    out.push({ t: isoHour(idx), v: round3(clip(truth[idx] + capacity * dev, 0, capacity)) });
  }
  return out;
}

// --- geometries (GeoJSON), from demo_data.py -------------------------------
type GeoJSON = Record<string, unknown>;
const poly = (ring: [number, number][]): GeoJSON => ({
  type: "Polygon",
  coordinates: [[...ring, ring[0]]],
});
const line = (pts: [number, number][]): GeoJSON => ({ type: "LineString", coordinates: pts });

export const OFFSHORE_AREA = poly([
  [12.893, 55.783], [12.92, 55.783], [12.92, 55.792], [12.893, 55.792],
]);
// DC link from the PV system (PV01) to the battery (B01) at Solar-Farm-1.
export const SOLAR_CABLE = line([
  [12.94151, 55.77887], [12.9435, 55.7805],
]);
export const OFFSHORE2_AREA = poly([
  [12.892, 55.8], [12.919, 55.8], [12.919, 55.81], [12.892, 55.81],
]);
export const SOLAR_FARM_AREA = poly([
  [12.937, 55.7765], [12.946, 55.7765], [12.946, 55.7815], [12.937, 55.7815],
]);
export const PV_ARRAY_AREA = poly([
  [12.9405, 55.7783], [12.9445, 55.7783], [12.9445, 55.7798], [12.9405, 55.7798],
]);

File: energydb_inspect/demo_data.py
"""Realistic, seeded synthetic time series for the demo (not energydb API, just
the numbers the notebook then `.write()`s). 72 hourly points starting 2026-01-01.

- solar  : daytime bell curve with cloud variability, 0 at night
- wind   : smooth random-walk capacity factor (lulls + gusts)
- speed  : wind speed correlated with the wind capacity factor
- battery: daily state-of-charge cycle + the (de)charge power that drives it
- forecast: revisions issued at advancing knowledge_times, each covering a
           bounded horizon window so it clearly starts and ends inside the
           timeline, fanning further from the truth the longer the lead time
"""

from __future__ import annotations

from datetime import UTC, datetime

import numpy as np
import pandas as pd
from shapely.geometry import LineString, Polygon

START = datetime(2026, 1, 1, tzinfo=UTC)
PERIODS = 72

# ── Geometry (lon, lat), real shapes the assets carry, shown on the map ──────
# Areas are polygons; turbines / PV system / battery stay points (passed as
# lat/lon). Children sit inside their parent's area. Both sites are in the Kalmar
# Sound region of southeast Sweden, ~16 km apart.
#
# Offshore wind farm in the Øresund: a compact lease-area polygon around the two
# turbines (T01 12.91331, T02 12.900, same latitude 55.78726).
OFFSHORE_AREA = Polygon(
    [(12.893, 55.783), (12.920, 55.783), (12.920, 55.792), (12.893, 55.792)]
)
# A second, similar-size wind farm a bit north, with a single turbine (T03 12.905/55.805).
OFFSHORE2_AREA = Polygon(
    [(12.892, 55.800), (12.919, 55.800), (12.919, 55.810), (12.892, 55.810)]
)
# Solar farm just east of the wind farm (centred on 12.94151/55.77887): the site
# boundary and, inside it, the rectangular footprint of the PV array.
SOLAR_FARM_AREA = Polygon(
    [(12.937, 55.7765), (12.946, 55.7765), (12.946, 55.7815), (12.937, 55.7815)]
)
PV_ARRAY_AREA = Polygon(
    [(12.9405, 55.7783), (12.9445, 55.7783), (12.9445, 55.7798), (12.9405, 55.7798)]
)
# DC link from the PV system (PV01 12.94151/55.77887) to the battery (B01 12.9435/55.7805).
SOLAR_CABLE = LineString([(12.94151, 55.77887), (12.9435, 55.7805)])


def _hours() -> pd.DatetimeIndex:
    return pd.date_range(START, periods=PERIODS, freq="1h")


def _frame(values: np.ndarray) -> pd.DataFrame:
    return pd.DataFrame({"valid_time": _hours(), "value": np.round(values, 3)})


def _hour_of_day() -> np.ndarray:
    return np.arange(PERIODS) % 24


def _wind_cf(seed: int = 2) -> np.ndarray:
    """Smooth capacity factor in [0, 1] via a mean-reverting random walk.

    High persistence + small steps keep the curve gentle (slow lulls and
    gusts) rather than jittery.
    """
    rng = np.random.default_rng(seed)
    cf = np.empty(PERIODS)
    cf[0] = 0.5
    for i in range(1, PERIODS):
        cf[i] = np.clip(0.92 * cf[i - 1] + 0.08 * 0.5 + rng.normal(0, 0.025), 0.0, 1.0)
    return cf


def wind_power(capacity: float = 3.5, seed: int = 2) -> pd.DataFrame:
    return _frame(capacity * _wind_cf(seed))


def wind_speed(seed: int = 2) -> pd.DataFrame:
    # m/s loosely implied by the capacity factor (cube-root-ish of a power curve)
    return _frame(3.0 + 9.0 * np.cbrt(_wind_cf(seed)))


def solar_power(capacity: float = 10.0, seed: int = 1) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    hod = _hour_of_day()
    bell = np.clip(np.sin(np.pi * (hod - 6) / 12), 0, None)  # up ~06:00, down ~18:00
    clouds = 0.8 + 0.2 * rng.random(PERIODS)
    return _frame(np.clip(capacity * bell * clouds, 0, capacity))


def battery_soc(seed: int = 3) -> pd.DataFrame:
    rng = np.random.default_rng(seed)
    hod = _hour_of_day()
    soc = 55 + 30 * np.sin(2 * np.pi * (hod - 15) / 24)  # charged by afternoon
    return _frame(np.clip(soc + rng.normal(0, 1.5, PERIODS), 5, 99))


def battery_power(seed: int = 3) -> pd.DataFrame:
    soc = battery_soc(seed)["value"].to_numpy()
    delta = np.diff(soc, prepend=soc[0])  # +charging / -discharging
    return _frame(delta * 5.0)  # %/h → MW (illustrative)


def cable_flow() -> pd.DataFrame:
    """Power across the PV-to-battery DC link: the battery's (de)charge power
    (positive while charging from the array, negative while discharging)."""
    return _frame(battery_power()["value"].to_numpy())


# Forecast revisions: (hours after START the forecast is issued, error scale).
# Six revisions issued every 8 h, each sharper than the last. Each covers a
# bounded FORECAST_HORIZON_H window from its issue time, so it starts and ends
# *inside* the 72 h timeline, making each revision's span easy to read.
FORECAST_HORIZON_H = 36
FORECAST_REVISIONS = [
    (0, 0.34),
    (8, 0.27),
    (16, 0.20),
    (24, 0.14),
    (32, 0.09),
    (40, 0.05),
]


def wind_power_forecast(
    capacity: float = 3.5,
    seed: int = 2,
    *,
    issued_at_h: int = 0,
    error: float = 0.15,
    horizon_h: int = FORECAST_HORIZON_H,
) -> pd.DataFrame:
    """A wind-power forecast issued ``issued_at_h`` hours after START.

    Covers a bounded ``horizon_h``-hour window from the issue time, so the
    revision clearly *starts* (peeling off the actual where it is issued) and
    *ends* (its horizon runs out) within the displayed timeline rather than
    running to the edge. It fans away from the truth smoothly as lead time
    grows, diverging most at the far end of the horizon. A smaller ``error``
    gives a later, sharper revision.
    """
    truth = capacity * _wind_cf(seed)
    end = min(issued_at_h + horizon_h, PERIODS)
    idx = np.arange(issued_at_h, end)
    lead = idx - issued_at_h
    growth = lead / max(horizon_h, 1)  # 0 at issue → 1 at the far end of the horizon
    # One slow, smooth departure wave (few wiggles) whose amplitude grows with
    # lead time; a per-issue phase/sign so the revisions look distinct.
    rng = np.random.default_rng(200 + issued_at_h)
    phase = rng.uniform(0, 2 * np.pi)
    sign = 1.0 if rng.random() < 0.5 else -1.0
    wave = np.sin(2 * np.pi * lead / 52.0 + phase)
    dev = sign * error * growth * wave
    fc = np.clip(truth[idx] + capacity * dev, 0, capacity)
    return pd.DataFrame({"valid_time": _hours()[idx], "value": np.round(fc, 3)})
